### PR TITLE
More dfhack.lua annotations

### PR DESCRIPTION
**Pull request by vallode**
Next round of annotations added to the core dfhack.lua file. Raised questions:

- I'm currently leaning towards abundant usage of `@nodiscard` which will warn users when a function has a return but they have not consumed it. Just something to keep in mind.
- with_finalize and with_onerror use variable arguments when the function they alias/call have known arguments, could be typed better.
- I think we should probably use aliases instead of new functions for the defclass overloads I tried to address this in #4445

### CODE REVIEW

File: library/lua/dfhack.lua [hunk 1]
```diff
@@ -74,10 +80,16 @@ end
 safecall = dfhack.safecall
 curry = dfhack.curry
 
+---@generic T
+---@param f fun(...):T
+---@param ... any
+---@return T
 function dfhack.pcall(f, ...)
```
Comment (myk002) [inline library/lua/dfhack.lua:87]: this one is tricky. On success, this function returns `true, ...` where `...` are the return values from `f` (different from the `@param ...`, which are the arguments to `f`).
On failure, it returns `false, err` where err is the error object, which I think is always string error message, at least whenever I've seen it used.

File: library/lua/dfhack.lua [hunk 2]
```diff
@@ -86,9 +98,12 @@ function qerror(msg, level)
     dfhack.error(msg, (level or 1) + 1, false)
 end
 
+---@param ... any
 function dfhack.with_finalize(...)
     return dfhack.call_with_finalizer(0,true,...)
 end
+
+---@param ... any
 function dfhack.with_onerror(...)
```
Comment (myk002) [inline library/lua/dfhack.lua:107]: I agree that these type signatures could be improved by defining the known first N args. The Lua docs define them, why aren't they defined here? https://docs.dfhack.org/en/latest/docs/dev/Lua%20API.html#locking-and-finalization

File: library/lua/dfhack.lua [hunk 3]
```diff
@@ -97,6 +112,11 @@ local function call_delete(obj)
     if obj then obj:delete() end
 end
 
+---@generic T
+---@param obj DFObject
+---@param fn fun(...):T
+---@param ...? any
+---@return T
```
Comment (myk002) [inline library/lua/dfhack.lua:119]: `T` can be multiple values. Is that adequately expressed with a single `T`?

File: library/lua/dfhack.lua [hunk 4]
```diff
@@ -215,10 +240,12 @@ local function print_element(k, v)
     dfhack.println(string.format("%-23s\t = %s", tostring(k), tostring(v)))
 end
 
+---@param table table
```
Comment (myk002) [inline library/lua/dfhack.lua:243]: could also be a c++ vector. do we want to draw that disctinction?

File: library/lua/dfhack.lua [hunk 5]
```diff
@@ -636,33 +688,54 @@ function dfhack.gui.getViewscreenByType(scr_type, n)
...
 local function persistent_getData(which, key, default)
     local serialized = dfhack.persistent['get'..which..'DataString'](key)
     if not serialized then return default end
     return require('json').decode(serialized) or default
 end
 
+---@param which string
+---@param key string
+---@param data any
 function persistent_saveData(which, key, data)
```
Comment (myk002) [inline library/lua/dfhack.lua:711]: `persistent_saveData` should be `local` like `persistent_getData` -- could you make that fix?

File: library/lua/dfhack.lua [hunk 6]
```diff
@@ -8,6 +8,8 @@
 
 ---@class dfhack
 ---@field BASE_G _G Original Lua global environment
+---@field is_core_context boolean
+---@field is_interactive fun(): boolean
```
[the remark this answers is not in the document]
Comment (vallode) [reply library/lua/dfhack.lua:12]: For now this is my duct tape method, it works but lacks some of the benefits a proper function decleration gives in LuaLS (there are a few issues open regarding these inline function fields).

I would leave it like this until we have a discussion/brainstorm on finalizing a method of parsing exposed C++ functions. The methods are probably:

- The above, just defining fields on a global `dfhack` object amongst others for the modules.
- Creating a separate definitions file that can be yanked out by the annotations library (the above, but a bit more verbose)
- Add some kind of special tag to the C++ files that can be easily parsed, essentially the above but inlining the definitions with the methods
- Someone whips up a magical bug-free method of parsing the C++ functions perfectly and detecting all exposed ones automatically :magic_wand: 

Again I think this is going to be a slightly more in-depth discussion but I put these in here for now just to have them for the file.

File: library/lua/dfhack.lua (hunk 1 above)
Comment (vallode) [reply library/lua/dfhack.lua:87]: Probably a case where we run against the edge of what LuaLS can actually represent. I think in this case we just do:

```
---@generic T
---@param f fun(...): T
---@param ... any -- LuaLS does not like typing variable arguments and especially not inside of a inline function :-(
---@return boolean
---@return T|string ... -- LuaLS will not correctly type variable return types as T but this is not a major drawback.
``

File: library/lua/dfhack.lua (hunk 3 above)
Comment (vallode) [reply library/lua/dfhack.lua:119]: Changed to variable return but as stated above LuaLS doesn't acknowledge that with generics. The rest of the values will be unknown.

File: library/lua/dfhack.lua (hunk 4 above)
Comment (vallode) [reply library/lua/dfhack.lua:243]: If there is no special behaviour there I think it's fine not to include, if it quacks like a table...

File: library/lua/dfhack.lua (hunk 1 above)
Comment (myk002) [reply library/lua/dfhack.lua]: lgtm

File: library/lua/dfhack.lua [hunk 7]
```diff
@@ -86,17 +99,36 @@ function qerror(msg, level)
...
 end
 
-function dfhack.with_finalize(...)
-    return dfhack.call_with_finalizer(0,true,...)
+---@generic T
+---@generic S
+---@param cleanup_fn fun(...): S
+---@param fn fun(...): T
+---@param ... any
+---@return T|S
```
Comment (myk002) [inline library/lua/dfhack.lua:107]: The return value from `cleanup_fn` is not propagated here, just the return values from `fun`

same with `with_onerror` below

Comment (vallode) [reply library/lua/dfhack.lua:107]: I misread the docs!

**Review by myk002 — APPROVED** [review]

**Pull request merged**